zkevm-circuits/src/rlp_decoder.rs

```diff
@@ -0,0 +1,3230 @@
+//! The rlp decoding transaction list circuit implementation.
+
+use std::marker::PhantomData;
+
+use crate::{
+    evm_circuit::util::{
+        constraint_builder::{BaseConstraintBuilder, ConstrainBuilderCommon},
+        rlc,
+    },
+    impl_expr,
+    rlp_decoder_tables::{
+        RlpDecodeRule, RlpDecoderFixedTable, RlpDecoderFixedTableTag, RLP_TX_FIELD_DECODE_RULES,
+    },
+    table::KeccakTable,
+    util::{log2_ceil, Challenges, SubCircuit, SubCircuitConfig},
+    witness,
+};
+use eth_types::{Field, Signature, Transaction, Word};
+use ethers_core::{types::TransactionRequest, utils::rlp};
+use gadgets::{
+    less_than::{LtChip, LtConfig, LtInstruction},
+    util::{and, not, or, sum},
+};
+use halo2_proofs::{
+    circuit::{Layouter, Region, SimpleFloorPlanner, Value},
+    plonk::{
+        Advice, Circuit, Column, ConstraintSystem, Error, Expression, Fixed, SecondPhase, Selector,
+    },
+    poly::Rotation,
+};
+use keccak256::plain::Keccak;
+
+use crate::util::Expr;
+pub use halo2_proofs::halo2curves::{
+    group::{
+        ff::{Field as GroupField, PrimeField},
+        prime::PrimeCurveAffine,
+        Curve, Group, GroupEncoding,
+    },
+    secp256k1::{self, Secp256k1Affine, Secp256k1Compressed},
+};
+use mock::MockTransaction;
+
+const NUM_BLINDING_ROWS: usize = 64;
+
+type RlpDecoderFixedTable6Columns = RlpDecoderFixedTable<6>;
+
+/// RlpDecodeTypeTag is used to index the flag of rlp decoding type
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub enum RlpDecodeTypeTag {
+    #[default]
+    /// Nothing for no rlp decoding
+    DoNothing = 0,
+    /// SingleByte: 0x00 - 0x7f
+    SingleByte,
+    /// NullValue: 0x80
+    NullValue,
+    /// ShortString: 0x81~0xb7, value means bytes without leading 0
+    ShortStringValue,
+    /// ShortString: 0x81~0xb7, bytes contains leading 0
+    ShortStringBytes,
+    /// LongString: 0xb8
+    LongString1,
+    /// LongString: 0xb9
+    LongString2,
+    /// LongString: 0xba
+    LongString3,
+    /// EmptyList: 0xc0
+    EmptyList,
+    /// ShortList: 0xc1 ~ 0xf7
+    ShortList,
+    /// LongList1: 0xf8
+    LongList1,
+    /// LongList2: 0xf9, 0xFFFF upto (64K)
+    LongList2,
+    /// LongList3: 0xfa, 0xFFFFFF upto (16M)
+    LongList3,
+    /// PartialRlp: for those rlp that is not complete
+    PartialRlp,
+    /// RlpDecodeTypeNum,
+    RlpDecodeTypeNum,
```

Comment:
What is the intention behind this value? I see it being used as a default value but there are no constraints using it.

Is it mean to be seen in the witness at all? If not, would it make sense to prohibit this value from appearing using constraints?

Reply:
Fixed, changed to a unified way of telling length/size of a enum.